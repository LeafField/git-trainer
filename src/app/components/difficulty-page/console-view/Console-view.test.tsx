import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ConsoleView from "./Console-view";

import * as Router from "next/navigation";

jest.mock("next/navigation");

const mockPush = jest.fn();

const mockCallback = jest.fn();

beforeEach(() => {
  jest.spyOn(Router, "useRouter").mockImplementation(() => ({
    back: jest.fn(),
    push: mockPush,
    forward: jest.fn(),
    refresh: jest.fn(),
    replace: jest.fn(),
    prefetch: jest.fn(),
  }));
});

afterEach(() => {
  jest.clearAllMocks();
});

describe("ConsoleViewの単体テスト", () => {
  it("ConsoleViewが正しくレンダリングされているか", async () => {
    render(
      <ConsoleView
        finished={false}
        question="リモートブランチをoriginに登録したい"
        answer="git remote add origin URL"
        nextCallback={mockCallback}
      />
    );
    expect(
      screen.getByText("リモートブランチをoriginに登録したい")
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "Git Empty" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("textbox", { name: "C:users/gitEmpty>" })
    ).toBeInTheDocument();
    expect(
      screen.queryByText("Would you like to start the next problem?(y/n)")
    ).not.toBeInTheDocument();
  });

  it("正常系:課題に正解した場合、次の課題に進むためのコールバック関数が発火するか", async () => {
    const { container } = render(
      <ConsoleView
        finished={false}
        question="リモートブランチをoriginに登録したい"
        answer="git remote add origin URL"
        nextCallback={mockCallback}
      />
    );
    await userEvent.type(
      screen.getByRole("textbox", { name: "C:users/gitEmpty>" }),
      "git remote add origin URL"
    );
    await userEvent.type(
      screen.getByRole("textbox", { name: "C:users/gitEmpty>" }),
      "{enter}"
    );
    waitFor(() => {
      expect(
        screen.getByText("Would you like to start the next problem?(y/n)")
      ).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });

    await userEvent.type(
      screen.getByRole("textbox", {
        name: "Would you like to start the next problem?(y/n)",
      }),
      "y"
    );

    await userEvent.type(
      screen.getByRole("textbox", {
        name: "Would you like to start the next problem?(y/n)",
      }),
      "{enter}"
    );
    waitFor(() => {
      expect(mockCallback).toHaveBeenCalled();
    });
  });

  it("正常系:問題に正解した上で、次の問題に進まない場合mockPushが発火するか", async () => {
    render(
      <ConsoleView
        finished={false}
        question="リモートブランチをoriginに登録したい"
        answer="git remote add origin URL"
        nextCallback={mockCallback}
      />
    );
    await userEvent.type(
      screen.getByRole("textbox", { name: "C:users/gitEmpty>" }),
      "git remote add origin URL"
    );
    await userEvent.type(
      screen.getByRole("textbox", { name: "C:users/gitEmpty>" }),
      "{enter}"
    );

    await userEvent.type(
      screen.getByRole("textbox", {
        name: "Would you like to start the next problem?(y/n)",
      }),
      "n"
    );

    await userEvent.type(
      screen.getByRole("textbox", {
        name: "Would you like to start the next problem?(y/n)",
      }),
      "{enter}"
    );
    waitFor(() => {
      expect(mockPush).toHaveBeenCalledWith("/");
    });
  });

  it("異常系:問題に正解した後の問いに(y/n)以外の回答をする", async () => {
    render(
      <ConsoleView
        question="リモートブランチをoriginに登録したい"
        answer="git remote add origin URL"
        nextCallback={mockCallback}
        finished={false}
      />
    );
    await userEvent.type(
      screen.getByRole("textbox", { name: "C:users/gitEmpty>" }),
      "git remote add origin URL"
    );
    await userEvent.type(
      screen.getByRole("textbox", { name: "C:users/gitEmpty>" }),
      "{enter}"
    );

    await userEvent.type(
      screen.getByRole("textbox", {
        name: "Would you like to start the next problem?(y/n)",
      }),
      "a"
    );

    await userEvent.type(
      screen.getByRole("textbox", {
        name: "Would you like to start the next problem?(y/n)",
      }),
      "{enter}"
    );
    waitFor(() => {
      expect(mockPush).not.toHaveBeenCalled();
      expect(mockCallback).not.toHaveBeenCalled();
      expect(
        screen.getByRole("textbox", {
          name: "Would you like to start the next problem?(y/n)",
        })
      ).toHaveValue("");
    });
  });

  it("異常系:問題に不正解の場合に、不正解のメッセージが表示されるか", async () => {
    const { container } = render(
      <ConsoleView
        question="リモートブランチをoriginに登録したい"
        answer="git remote add origin URL"
        nextCallback={mockCallback}
        finished={false}
      />
    );

    await userEvent.type(
      screen.getByRole("textbox", { name: "C:users/gitEmpty>" }),
      "git"
    );

    await userEvent.type(
      screen.getByRole("textbox", { name: "C:users/gitEmpty>" }),
      "{enter}"
    );
    waitFor(() => {
      expect(
        screen.getByText("不正解です！右上の検索バーから正解を検索してみてね！")
      ).toBeInTheDocument();
      expect(screen.queryByText("git")).not.toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  it("正常系:全問正解後にメッセージが流れ、router.pushが発火するか", async () => {
    jest.useFakeTimers();
    const { container } = render(
      <ConsoleView
        finished={true}
        question="リモートブランチをoriginに登録したい"
        answer="git remote add origin URL"
        nextCallback={mockCallback}
      />
    );
    jest.runAllTimers();
    expect(screen.getByText("Congratulations!")).toBeInTheDocument();
    expect(mockPush).toHaveBeenCalledWith("/difficulty");
    expect(container).toMatchSnapshot();
  });
});
