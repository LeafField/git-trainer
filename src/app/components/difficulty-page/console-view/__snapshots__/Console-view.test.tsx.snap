// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsoleViewの単体テスト 正常系:全問正解後にメッセージが流れ、router.pushが発火するか 1`] = `
<div>
  <main
    class="row-start-2 h-[29.125rem] max-w-[62.5rem] cursor-text overflow-hidden rounded-lg border border-consoleHeader text-[rgb(204,204,204)] lg:col-span-4 lg:col-start-2"
  >
    <h2
      class="h-6 bg-consoleHeader pl-3 text-base text-white"
    >
      Git Empty
    </h2>
    <div
      class="min-h-full bg-console px-3 pt-7"
    >
      <p>
        リモートブランチをoriginに登録したい
      </p>
      <div
        class="pt-4"
      >
        <p
          class="text-xl font-bold"
        >
          Congratulations!
        </p>
        <p
          class="pt-2"
        >
          ここまで遊んでくれてありがとうございます
        </p>
        <p>
          是非他の難易度も遊んでみてください
        </p>
      </div>
    </div>
  </main>
</div>
`;

exports[`ConsoleViewの単体テスト 正常系:課題に正解した場合、次の課題に進むためのコールバック関数が発火するか 1`] = `
<div>
  <main
    class="row-start-2 h-[29.125rem] max-w-[62.5rem] cursor-text overflow-hidden rounded-lg border border-consoleHeader text-[rgb(204,204,204)] lg:col-span-4 lg:col-start-2"
  >
    <h2
      class="h-6 bg-consoleHeader pl-3 text-base text-white"
    >
      Git Empty
    </h2>
    <div
      class="min-h-full bg-console px-3 pt-7"
    >
      <p>
        リモートブランチをoriginに登録したい
      </p>
      <form
        class="mt-6 flex max-w-full flex-col gap-1 sm:flex-row"
      >
        <label
          for="console"
        >
          C:users/gitEmpty&gt;
        </label>
        <input
          class="flex-1 bg-transparent outline-none"
          id="console"
          name="console"
          type="text"
        />
      </form>
      <div
        class="pt-4"
      >
        <p>
          Congratulations, you are correct!
        </p>
        <form
          class="flex max-w-full flex-col sm:flex-row"
        >
          <label
            for="consoleInput"
          >
            Would you like to start the next problem?
            (y/n)
          </label>
          <input
            class="ml-1 bg-transparent outline-none"
            id="consoleInput"
            type="text"
          />
        </form>
      </div>
    </div>
  </main>
</div>
`;

exports[`ConsoleViewの単体テスト 異常系:問題に不正解の場合に、不正解のメッセージが表示されるか 1`] = `
<div>
  <main
    class="row-start-2 h-[29.125rem] max-w-[62.5rem] cursor-text overflow-hidden rounded-lg border border-consoleHeader text-[rgb(204,204,204)] lg:col-span-4 lg:col-start-2"
  >
    <h2
      class="h-6 bg-consoleHeader pl-3 text-base text-white"
    >
      Git Empty
    </h2>
    <div
      class="min-h-full bg-console px-3 pt-7"
    >
      <p>
        リモートブランチをoriginに登録したい
      </p>
      <div
        class="pt-4"
      >
        不正解です！右上の検索バーから正解を検索してみてね！
      </div>
      <form
        class="mt-6 flex max-w-full flex-col gap-1 sm:flex-row"
      >
        <label
          for="console"
        >
          C:users/gitEmpty&gt;
        </label>
        <input
          class="flex-1 bg-transparent outline-none"
          id="console"
          name="console"
          type="text"
        />
      </form>
    </div>
  </main>
</div>
`;
